internal/pkg/source/source.go

```diff
@@ -62,54 +60,64 @@ func (s *Source) Pos() token.Pos {
 // New constructs a Source
 func New(in ssa.Node, config classifier) *Source {
 	s := &Source{
-		node:            in,
-		marked:          make(map[ssa.Node]bool),
-		config:          config,
-		maxInstrReached: map[*ssa.BasicBlock]int{},
+		node:   in,
+		marked: make(map[ssa.Node]bool),
+		config: config,
 	}
-	s.dfs(in)
+	s.dfs(in, map[*ssa.BasicBlock]int{}, nil)
 	return s
 }
 
 // dfs performs Depth-First-Search on the def-use graph of the input Source.
 // While traversing the graph we also look for potential sanitizers of this Source.
 // If the Source passes through a sanitizer, dfs does not continue through that Node.
-func (s *Source) dfs(n ssa.Node) {
+func (s *Source) dfs(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
 	if s.marked[n] {
 		return
 	}
+	mirCopy := map[*ssa.BasicBlock]int{}
+	for m, i := range maxInstrReached {
+		mirCopy[m] = i
+	}
 
 	if instr, ok := n.(ssa.Instruction); ok {
 		if !s.reachableFromSource(instr) {
 			return
 		}
-		s.record(instr)
+		// If the referrer is in a different block from the one we last visited,
+		// and it can't be reached from the block we are visiting, then stop visiting.
+		if ib := instr.Block(); lastBlockVisited != nil &&
+			ib != lastBlockVisited &&
+			!s.canReach(lastBlockVisited, ib) {
+			return
+		}
+		s.record(instr, mirCopy, &lastBlockVisited)
 	}
 	s.preOrder = append(s.preOrder, n)
 	s.marked[n] = true
 
-	s.visitReferrers(n)
+	s.visitReferrers(n, mirCopy, lastBlockVisited)
 
 	operands := n.Operands(nil)
 	if operands != nil {
-		s.visitOperands(n, operands)
+		s.visitOperands(n, operands, mirCopy, lastBlockVisited)
 	}
 }
 
-func (s *Source) record(target ssa.Instruction) {
+func (s *Source) record(target ssa.Instruction, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited **ssa.BasicBlock) {
 	b := target.Block()
-	s.lastBlockVisited = b
+	*lastBlockVisited = b
```

Comment:
Updating via a mutable `**ssa.BasicBlock` is a severe code smell to me.  Consider returning `(instrBlock *ssa.BasicBlock)` instead, or extracting `thisBlock = instr.Block()` to the calling function.

Reply:
Hmm. I agree that this smells pretty bad.

In fact, re-reading the code now, it seems to me that this does not need to be a method, and the code can be simplified enough to be inlined:
```go
b := instr.Block()
if i, ok := indexInBlock(instr); ok && mirCopy[b] < i {
	mirCopy[b] = i
}
lastBlockVisited = b
```